### ISSUE TITLE
conduit-proxy pegging cpu

### ISSUE THREAD

**Issue by siggy** [runconduit/conduit#519]
Building and running Conduit from master, I observe the conduit-proxy in the Grafana pod pegging the CPU, about a minute after startup.

Grafana is complaining about not being able to connect to grafana.com, which is a manifestation of #155. Confirmed this is the cause of the CPU pegging.

## Workarounds

We can work around this via one of two ways.

### Ignore outbound 443

Ignore port 443 on the outbound Grafana proxy-init container:
```yaml
- --outbound-ports-to-ignore
- "443"
```

### Disable update check

Disable the Grafana update check in `grafana.ini`:
```ini
[analytics]
check_for_updates = false
```

## Repro

```bash
DOCKER_TRACE=1 bin/docker-build
bin/conduit install --proxy-log-level debug,conduit_proxy=debug | kubectl apply -f -
```

Note that curl'ing Prometheus from the Grafana container hangs:

```bash
$ kubectl -n conduit exec -it grafana-68954c7876-txnsd /bin/bash -c grafana
$ curl -v prometheus.conduit.svc.cluster.local:9090
* Rebuilt URL to: prometheus.conduit.svc.cluster.local:9090/
* Hostname was NOT found in DNS cache
*   Trying 10.104.147.185...
* Connected to prometheus.conduit.svc.cluster.local (10.104.147.185) port 9090 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.38.0
> Host: prometheus.conduit.svc.cluster.local:9090
> Accept: */*
>

(hangs)
```

## CPU

```bash
$ docker stats $(docker ps --format={{.Names}})

CONTAINER ID        NAME                                                                                                                    CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
...
dc0e821e277f        k8s_conduit-proxy_grafana-68954c7876-txnsd_conduit_13a09de6-2184-11e8-b369-025000000001_0                               95.08%              984KiB / 11.71GiB     0.01%               0B / 0B             0B / 0B             2
...
```

### Grafana

```bash
$ kubectl -n conduit logs -f  $(kubectl --namespace=conduit get po --selector=conduit.io/control-plane-component=grafana -o jsonpath='{.items[*].metadata.name}') -c grafana
...
t=2018-03-06T22:49:39+0000 lvl=dbug msg="Failed to get plugins repo from grafana.net, Get https://grafana.com/api/plugins/versioncheck?slugIn=&grafanaVersion=5.0.0-beta4: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)"
```

### conduit-proxy

```bash
$ kubectl -n conduit logs -f  $(kubectl --namespace=conduit get po --selector=conduit.io/control-plane-component=grafana -o jsonpath='{.items[*].metadata.name}') -c conduit-proxy
DBUG tokio_core::reactor loop poll - Duration { secs: 10, nanos: 2779700 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57030, tv_nsec: 420110808 }
DBUG tokio_core::reactor notifying a task handle
DBUG tokio_core::reactor loop process - 0 events, Duration { secs: 0, nanos: 54500 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 12900 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57030, tv_nsec: 420184208 }
DBUG tokio_core::reactor "controller-client", set a timeout: 0
DBUG tokio_core::reactor "controller-client", added a timeout: 3
DBUG tokio_core::reactor "controller-client", updating a timeout: 3
DBUG tokio_core::reactor "controller-client", updating a timeout: 0
DBUG tokio_core::reactor "controller-client", scheduling direction for: 0
DBUG tokio_core::reactor "controller-client", blocking
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 3084100 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 64300 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57030, tv_nsec: 423350508 }
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), scheduling direction for: 2
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), blocking
DBUG h2::codec::framed_write ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), send; frame=Frame::Headers(Headers { stream_id: StreamId(11), stream_dep: None, flags: HeadersFlag { end_stream: false, end_headers: true, padded: false, priority: false } })
DBUG tokio_core::reactor loop process - 2 events, Duration { secs: 0, nanos: 3792400 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 12800 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57030, tv_nsec: 427186708 }
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), scheduling direction for: 2
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), blocking
DBUG h2::codec::framed_write ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), send; frame=Frame::Data(Data { stream_id: StreamId(11), flags: DataFlags { end_stream: true }, pad_len: None })
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 3693400 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 13000 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57030, tv_nsec: 431057108 }
DBUG tokio_core::reactor notifying a task handle
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 15300 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 9200 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57030, tv_nsec: 431089508 }
DBUG h2::codec::framed_read ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), received; frame=Frame::Headers(Headers { stream_id: StreamId(11), stream_dep: None, flags: HeadersFlag { end_stream: false, end_headers: true, padded: false, priority: false } })
DBUG h2::codec::framed_read ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), received; frame=Frame::Data(Data { stream_id: StreamId(11), flags: DataFlags, pad_len: None })
DBUG h2::codec::framed_read ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), received; frame=Frame::Headers(Headers { stream_id: StreamId(11), stream_dep: None, flags: HeadersFlag { end_stream: true, end_headers: true, padded: false, priority: false } })
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), scheduling direction for: 2
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), blocking
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 1272200 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 15000 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57030, tv_nsec: 432384408 }
DBUG tokio_core::reactor::timeout_token "controller-client", cancel timeout 3
DBUG tokio_core::reactor "controller-client", cancel a timeout: 3
DBUG tokio_core::reactor "controller-client", updating a timeout: 0
DBUG tokio_core::reactor "controller-client", scheduling direction for: 0
DBUG tokio_core::reactor "controller-client", blocking
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 101900 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 9500 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57030, tv_nsec: 432500508 }
DBUG tokio_core::reactor loop process - 0 events, Duration { secs: 0, nanos: 9800 }
DBUG tokio_core::reactor loop poll - Duration { secs: 65, nanos: 11767500 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57037, tv_nsec: 544646508 }
DBUG tokio_core::reactor notifying a task handle
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 351500 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 91300 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57037, tv_nsec: 545191508 }
DBUG tokio_core::reactor scheduling direction for: 2
DBUG tokio_core::reactor blocking
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 741600 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 12000 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57037, tv_nsec: 546367708 }
DBUG tokio_core::reactor notifying a task handle
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 552200 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 17400 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57037, tv_nsec: 547072608 }
DBUG tokio_core::reactor loop poll - Duration { secs: 9, nanos: 997204200 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57040, tv_nsec: 429721508 }
DBUG tokio_core::reactor notifying a task handle
DBUG tokio_core::reactor loop process - 0 events, Duration { secs: 0, nanos: 49700 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 12000 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57040, tv_nsec: 429789108 }
DBUG tokio_core::reactor "controller-client", set a timeout: 0
DBUG tokio_core::reactor "controller-client", added a timeout: 3
DBUG tokio_core::reactor "controller-client", updating a timeout: 3
DBUG tokio_core::reactor "controller-client", updating a timeout: 0
DBUG tokio_core::reactor "controller-client", scheduling direction for: 0
DBUG tokio_core::reactor "controller-client", blocking
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 88500 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 7000 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57040, tv_nsec: 429889408 }
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), scheduling direction for: 2
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), blocking
DBUG h2::codec::framed_write ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), send; frame=Frame::Headers(Headers { stream_id: StreamId(13), stream_dep: None, flags: HeadersFlag { end_stream: false, end_headers: true, padded: false, priority: false } })
DBUG tokio_core::reactor loop process - 2 events, Duration { secs: 0, nanos: 123400 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 7100 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57040, tv_nsec: 430027108 }
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), scheduling direction for: 2
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), blocking
DBUG h2::codec::framed_write ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), send; frame=Frame::Data(Data { stream_id: StreamId(13), flags: DataFlags { end_stream: true }, pad_len: None })
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 47000 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 4319500 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57040, tv_nsec: 434473108 }
DBUG tokio_core::reactor loop process - 0 events, Duration { secs: 0, nanos: 22100 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 3044900 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57040, tv_nsec: 437545408 }
DBUG tokio_core::reactor notifying a task handle
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 41900 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 65500 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57040, tv_nsec: 437660908 }
DBUG h2::codec::framed_read ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), received; frame=Frame::Headers(Headers { stream_id: StreamId(13), stream_dep: None, flags: HeadersFlag { end_stream: false, end_headers: true, padded: false, priority: false } })
DBUG h2::codec::framed_read ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), received; frame=Frame::Data(Data { stream_id: StreamId(13), flags: DataFlags, pad_len: None })
DBUG h2::codec::framed_read ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), received; frame=Frame::Headers(Headers { stream_id: StreamId(13), stream_dep: None, flags: HeadersFlag { end_stream: true, end_headers: true, padded: false, priority: false } })
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), scheduling direction for: 2
DBUG tokio_core::reactor ("controller-client", "HostAndPort { host: DnsName(\"proxy-api.conduit.svc.cluster.local\"), port: 8086 }"), blocking
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 893300 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 849300 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57040, tv_nsec: 439437508 }
DBUG tokio_core::reactor::timeout_token "controller-client", cancel timeout 3
DBUG tokio_core::reactor "controller-client", cancel a timeout: 3
DBUG tokio_core::reactor "controller-client", updating a timeout: 0
DBUG tokio_core::reactor "controller-client", scheduling direction for: 0
DBUG tokio_core::reactor "controller-client", blocking
DBUG tokio_core::reactor loop process - 1 events, Duration { secs: 0, nanos: 139100 }
DBUG tokio_core::reactor loop poll - Duration { secs: 0, nanos: 9000 }
DBUG tokio_core::reactor loop time - Instant { tv_sec: 57040, tv_nsec: 439595408 }
DBUG tokio_core::reactor loop process - 0 events, Duration { secs: 0, nanos: 8200 }
```

**Comment by briansmith** [comment]
If the cause is #155 then let's fix #155. I am working towards that now.

> Disable the Grafana update check in grafana.ini:
> ```
> [analytics]
> check_for_updates = false
> ```

Maybe we should do this anyway?

**Comment by siggy** [comment]
Review to disable update check: #521.

**Comment by briansmith** [comment]
@siggy wrote (elsewhere):
> @briansmith i believe #519 goes away when #155 is fixed...
> that said, perhaps it's a separate issue that the proxy pegs the cpu when it's given an address it's unable to route?

Yes, we need to find out why the proxy has gone haywire in this situation.

**Comment by olix0r** [comment]
marking this P0 to at least develop a working theory of this issue so that we can assess impact.

**Comment by hawkw** [comment]
@siggy and I found the bug that was causing this issue and fixed it in e5a7607c3be3f1001b196c5a83b9ee88b51a9775.

 I'd like to try and put together a unit test against a regression here, but would be happy to go ahead and open a PR for review.
